prevent-broker-scale-down-if-contains-partition .md

```diff
@@ -0,0 +1,57 @@
+So this check will target the broker which are going to be removed i.e. the pods with the highest number.
+
+Methods to be created in the process:
+
+- `getTopicNamesAndDescriptions` (Boolean Type) - This method will be placed inside the `scaleDown()` method in `KafkaReconciler` class and will provide us with the Collection of `TopicDescription`.
+- `describeTopics()` and `topicNames()`  - These two methods will provide us with the description and names of all the topics ie first we will get the names of the topics using the `topicNames` method and using these names we will extract the description of the topics.
+- `brokerHasAnyTopics()` - This method will take as input a podId and a `Collection<TopicDescription>`.
+  It will go through all the broker and see if any partitions are assigned to it.
+
+### Flow :
+
+When the replicas are changed inside the Kafka resource, reconciliation will happen.
+During this reconciliation, we move to the `scaleDown()` method present in the `KafkaReconciler` class.
+The `scaleDown()` method contains the complete mechanism on how the broker is scaled down. So there, we will first create our `adminClient` and provide it with the Secrets.
+Then we will call the `getTopicNamesAndDescriptions()` method which will give us the `Collection<TopicDescription>`. 
+
+The `getTopicNamesAndDescriptions()` method will call the `topicNames()` method to get the set of topic names and then these topic names will be used by the `describeTopics()` methods to get the `Collection<TopicDescription>`.
+Once the `Collection<TopicDescription>` is computed we will move back to the `scaledown()` method and there we will run a loop from `(previous replicas - 1)` to `current replicas` and inside the loop we will call the `brokersHasAnyTopics()` method which will take the `Collection<TopicDescription>` and `podId(index of the loop)` to check if the broker contains any partition and provide us with a boolean result. 
+
+### What to do if a broker contains partitions?
+If a partition is detected over a broker then we should trigger some log warnings and let the partitions remain the same as they were before.
+For example, if we try to scale down the broker from 5 to 3, but the check detected that the broker to be scaled down contains a partition then we should keep the replicas as 5 only display the appropriate warnings.
+
+To handle the case, if broker contain some partitions, we can make use of an `if` statement. If we see the current code in the `scaleDown()` method we can see that we are checking whether we are using `StatefulSets` or `StrimziPodSets` and then doing the scale down accordingly, so we can introduce an `if` just before this logic. This `if` condition will check if the boolean result was `true` or `false`. If the result is `false` then we will return a succeeded future i.e. no need to scale down but if the result is `true` then we can run the logic of scale down as per `StatefulSets` or `StrimziPodSets`.
```

Comment:
I don't understand this paragraph. If what result is false? What is this if statement checking? Weather we are using STSs or SPSs? Or if the broker has partitions?

Reply:
The `if` condition will check the boolean result stored in a variable w.r.t to whether the broker had partitions or not. If the result comes out to be `True` we will just return a succeeded future and skip the other logic `STS` and `SPS` since we don't need to scale down at all

Reply:
I think this comment is related to my more general comment where I am saying that the proposal sounds to describe too many programming constructs than describing the overall solution. I found this paragraph hard to understand too.

Reply:
Yeah the proposal should stick to high level processes and discriptions and not talk about implementation.